File: module-services/src/main/java/eu/geekhome/services/automation/StateDeviceAutomationUnit.java
package eu.geekhome.services.automation;


import eu.geekhome.services.hardware.Port;
import eu.geekhome.services.hardware.Relay;

import java.util.Map;

public abstract class StateDeviceAutomationUnit implements IDeviceAutomationUnit<State>, IMultistateDeviceAutomationUnit {
    public State _currentState;
    public ControlMode _controlMode;

    private final Map<String, State> _states;

    public StateDeviceAutomationUnit(Map<String, State> states, String initialState) {
        _states = states;
        _currentState = _states.get(initialState);
        _controlMode = ControlMode.Auto;
    }

    @Override
    public void setControlMode(ControlMode value) {
        _controlMode = value;
    }

    @Override
    public ControlMode getControlMode() {
        return _controlMode;
    }

    @Override
    public State getValue() {
        return _currentState;
    }

    protected void setCurrentState(String currentState) {
        _currentState = _states.get(currentState);
    }

    @Override
    public void changeState(String state, ControlMode controlMode, String code, String actor) {
        if (!_currentState.getName().getId().equals(state) || controlMode != getControlMode()) {
            setCurrentState(state);
            setControlMode(controlMode);
        }
    }

    public  void changeStateInternal(String state, ControlMode controlMode) throws Exception {
        changeState(state, controlMode, null, "SYSTEM");
    }

    @Override
    public EvaluationResult buildEvaluationResult() {
        return new EvaluationResult(getValue(), _currentState.getName(), isSignaled(), null, getControlMode());
    }

    protected boolean isSignaled() {
        return _currentState.isSignaled();
    }

    protected String getCurrentStateId() {
        return _currentState.getName().getId();
    }

    protected static <T> void changeOutputPortStateIfNeeded(Port<Relay> port, Relay state, boolean invalidate) throws Exception {
        if (port != null && state != null && (invalidate || !state.equals(port.read()))) {
            port.write(state);
        }
    }

    protected static <T> void changeOutputPortStateIfNeeded(Port<Relay> port, Relay state) throws Exception {
        changeOutputPortStateIfNeeded(port, state, false);
    }
}

File: module-services/src/main/java/eu/geekhome/services/automation/StateDeviceAutomationUnit.kt
package eu.geekhome.services.automation

import eu.geekhome.services.hardware.Port
import eu.geekhome.services.hardware.Relay

abstract class StateDeviceAutomationUnit(
    private val states: Map<String, State>,
    initialState: String) :
    DeviceAutomationUnit<State>(buildEvaluationResult(states[initialState]!!, ControlMode.Auto)), IStateDeviceAutomationUnit {

    var currentState: State
    override var controlMode: ControlMode = ControlMode.Auto

    protected fun setCurrentState(stateId: String) {
        this.currentState = states[stateId]!!
    }

    override fun changeState(state: String, controlMode: ControlMode, code: String?, actor: String?) {
        if (currentState.id != state || controlMode != controlMode) {
            setCurrentState(state)
            this.controlMode = controlMode
            configureNewState(state)

            lastEvaluation = buildEvaluationResult(currentState, controlMode)
        }
    }

    abstract fun configureNewState(state: String)

    companion object {
        @Throws(Exception::class)
        @JvmStatic
        protected fun <T> changeOutputPortStateIfNeeded(
            port: Port<Relay>?,
            state: Relay?,
            invalidate: Boolean = false
        ) {
            if (port != null && state != null && (invalidate || !state.equals(port.read()))) {
                port.write(state)
            }
        }

        fun buildEvaluationResult(state: State, controlMode: ControlMode) : EvaluationResult<State> {
            return EvaluationResult(
                interfaceValue = state.name,
                value = state,
                isSignaled = state.isSignaled,
                controlMode = controlMode
            )
        }
    }

    init {
        currentState = states[initialState]!!
    }
}